src/main/scala/nl.knaw.dans.easy.deposit/WorkerThread.scala

```diff
@@ -0,0 +1,136 @@
+/**
+ * Copyright (C) 2018 DANS - Data Archiving and Networked Services (info@dans.knaw.nl)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package nl.knaw.dans.easy.deposit
+
+import java.nio.file.StandardWatchEventKinds.{ ENTRY_CREATE, ENTRY_DELETE }
+import java.nio.file.{ WatchEvent, WatchKey, WatchService }
+
+import better.files.File
+import nl.knaw.dans.easy.deposit.docs.StateInfo.State
+import nl.knaw.dans.lib.error._
+import nl.knaw.dans.lib.logging.DebugEnhancedLogging
+
+import scala.util.Try
+
+object WorkerThread extends DebugEnhancedLogging {
+
+  val stagedBaseDir = File("target") // TODO allow inject/config to unit-test takeLoop
+
+  def main(args: Array[String]): Unit = {
+
+    // TODO create watchService as first action in EasyDepositApiService.start
+    //  close it as last action in EasyDepositApiService.stop
+    stagedBaseDir.watchService.apply { watchService =>
+      stagedBaseDir.register(watchService, Seq(ENTRY_CREATE, ENTRY_DELETE))
+      handleWatchService(watchService) // TODO in a sub-thread created by EasyDepositApiService.start
+    }
+  }
+
+  def handleWatchService(watchService: WatchService): Unit = {
+    // https://howtodoinjava.com/java8/java-8-watchservice-api-tutorial/
+    // take, because polling skips staged-upload-dirs created and deleted within a single cycle
+    handleAbortedByHardShutdown()
+    Iterator.continually(()).foreach { _ =>
+      Try(watchService.take())
```

Comment:
@rvanheest did you try `take` in the past or `poll`? If the first I hope you remember the problems. Note also my comments in `handleWatchKey`, I don't understand every detail there.

Reply:
I remember using the `watchService` indirectly in combination with RxJava. See their [implementation](https://github.com/ReactiveX/RxJavaFileUtils/blob/3ee8d12fc2947eb5d7e65cfb13bfd6ea13be884e/src/main/java/rx/fileutils/FileSystemWatcher.java#L72-L77) for some inspiration. I think you should use `register` for 'push-based' behavior, rather than pulling continuously...

Reply:
Looking somewhat further, though, I see they use `take` as well... https://github.com/ReactiveX/RxJavaFileUtils/blob/3ee8d12fc2947eb5d7e65cfb13bfd6ea13be884e/src/main/java/rx/fileutils/FileSystemWatcher.java#L102-L124